File: Overlord.Core/Overlord.Core/UpgradeSystem.cs
using Overlord.Core.Models;

namespace Overlord.Core;

/// <summary>
/// Platform-agnostic upgrade system for fleet-wide weapon research.
/// Manages weapon tier progression (Laser → Missile → Photon Torpedo).
/// </summary>
public class UpgradeSystem
{
    private readonly GameState _gameState;

    /// <summary>
    /// Event fired when weapon research starts.
    /// Parameters: (faction, weaponTier)
    /// </summary>
    public event Action<FactionType, WeaponTier>? OnResearchStarted;

    /// <summary>
    /// Event fired when weapon research completes.
    /// Parameters: (faction, weaponTier)
    /// </summary>
    public event Action<FactionType, WeaponTier>? OnResearchCompleted;

    // Research state per faction
    private WeaponTier _playerWeaponTier = WeaponTier.Laser;
    private WeaponTier _aiWeaponTier = WeaponTier.Laser;

    private WeaponTier? _playerResearchingTier = null;
    private WeaponTier? _aiResearchingTier = null;

    private int _playerResearchTurnsRemaining = 0;
    private int _aiResearchTurnsRemaining = 0;

    public UpgradeSystem(GameState gameState)
    {
        _gameState = gameState ?? throw new ArgumentNullException(nameof(gameState));
    }

    /// <summary>
    /// Gets the current weapon tier for a faction.
    /// </summary>
    public WeaponTier GetWeaponTier(FactionType faction)
    {
        return faction == FactionType.Player ? _playerWeaponTier : _aiWeaponTier;
    }

    /// <summary>
    /// Checks if a faction is currently researching a weapon upgrade.
    /// </summary>
    public bool IsResearching(FactionType faction)
    {
        return faction == FactionType.Player ? _playerResearchingTier.HasValue : _aiResearchingTier.HasValue;
    }

    /// <summary>
    /// Gets the weapon tier currently being researched (null if none).
    /// </summary>
    public WeaponTier? GetResearchingTier(FactionType faction)
    {
        return faction == FactionType.Player ? _playerResearchingTier : _aiResearchingTier;
    }

    /// <summary>
    /// Gets the remaining research turns for a faction.
    /// </summary>
    public int GetResearchTurnsRemaining(FactionType faction)
    {
        return faction == FactionType.Player ? _playerResearchTurnsRemaining : _aiResearchTurnsRemaining;
    }

    /// <summary>
    /// Starts weapon upgrade research for a faction.
    /// </summary>
    /// <param name="faction">Faction researching</param>
    /// <returns>True if research started, false if failed</returns>
    public bool StartResearch(FactionType faction)
    {
        // Get current and target tiers
        var currentTier = GetWeaponTier(faction);
        var nextTier = GetNextTier(currentTier);

        // Cannot research if already at max tier
        if (nextTier == null)
            return false; // Already at Photon Torpedo

        // Cannot research if already researching
        if (IsResearching(faction))
            return false; // Research already in progress

        // Get cost
        int cost = UpgradeCosts.GetResearchCost(nextTier.Value);

        // Validate resources (from faction resources)
        var factionState = faction == FactionType.Player ? _gameState.PlayerFaction : _gameState.AIFaction;
        if (factionState.Resources.Credits < cost)
            return false; // Insufficient credits

        // Deduct cost
        factionState.Resources.Credits -= cost;

        // Start research
        int researchTime = UpgradeCosts.GetResearchTime(nextTier.Value);

        if (faction == FactionType.Player)
        {
            _playerResearchingTier = nextTier;
            _playerResearchTurnsRemaining = researchTime;
        }
        else
        {
            _aiResearchingTier = nextTier;
            _aiResearchTurnsRemaining = researchTime;
        }

        OnResearchStarted?.Invoke(faction, nextTier.Value);
        return true;
    }

    /// <summary>
    /// Updates research progress (called each turn during Income Phase).
    /// </summary>
    public void UpdateResearch()
    {
        // Update player research
        if (_playerResearchingTier.HasValue && _playerResearchTurnsRemaining > 0)
        {
            _playerResearchTurnsRemaining--;

            if (_playerResearchTurnsRemaining <= 0)
            {
                CompleteResearch(FactionType.Player);
            }
        }

        // Update AI research
        if (_aiResearchingTier.HasValue && _aiResearchTurnsRemaining > 0)
        {
            _aiResearchTurnsRemaining--;

            if (_aiResearchTurnsRemaining <= 0)
            {
                CompleteResearch(FactionType.AI);
            }
        }
    }

    /// <summary>
    /// Completes weapon research and upgrades all Battle Cruisers.
    /// </summary>
    private void CompleteResearch(FactionType faction)
    {
        WeaponTier? researchedTier = faction == FactionType.Player ? _playerResearchingTier : _aiResearchingTier;

        if (!researchedTier.HasValue)
            return;

        // Upgrade faction weapon tier
        if (faction == FactionType.Player)
        {
            _playerWeaponTier = researchedTier.Value;
            _playerResearchingTier = null;
            _playerResearchTurnsRemaining = 0;
        }
        else
        {
            _aiWeaponTier = researchedTier.Value;
            _aiResearchingTier = null;
            _aiResearchTurnsRemaining = 0;
        }

        // Upgrade all existing Battle Cruisers owned by this faction
        var battleCruisers = _gameState.Craft
            .Where(c => c.Owner == faction && c.Type == CraftType.BattleCruiser)
            .ToList();

        foreach (var craft in battleCruisers)
        {
            // Apply weapon upgrade (future combat system will use this)
            // For now, we can just track the faction's weapon tier
        }

        OnResearchCompleted?.Invoke(faction, researchedTier.Value);
    }

    /// <summary>
    /// Gets the next weapon tier in progression.
    /// </summary>
    private WeaponTier? GetNextTier(WeaponTier current)
    {
        return current switch
        {
            WeaponTier.Laser => WeaponTier.Missile,
            WeaponTier.Missile => WeaponTier.PhotonTorpedo,
            WeaponTier.PhotonTorpedo => null, // Max tier
            _ => null
        };
    }

    /// <summary>
    /// Checks if a faction can afford to start the next weapon research.
    /// </summary>
    public bool CanAffordNextResearch(FactionType faction)
    {
        var currentTier = GetWeaponTier(faction);
        var nextTier = GetNextTier(currentTier);

        if (nextTier == null)
            return false; // Already at max tier

        int cost = UpgradeCosts.GetResearchCost(nextTier.Value);
        var factionState = faction == FactionType.Player ? _gameState.PlayerFaction : _gameState.AIFaction;

        return factionState.Resources.Credits >= cost;
    }

    /// <summary>
    /// Gets the damage modifier for a faction's current weapon tier.
    /// </summary>
    public float GetDamageModifier(FactionType faction)
    {
        var tier = GetWeaponTier(faction);
        return UpgradeCosts.GetDamageModifier(tier);
    }
}

File: Overlord.Phaser/src/core/UpgradeSystem.ts
import { GameState } from './GameState';
import { FactionType, WeaponTier, CraftType } from './models/Enums';
import { UpgradeCosts } from './models/UpgradeModels';

/**
 * Platform-agnostic upgrade system for fleet-wide weapon research.
 * Manages weapon tier progression (Laser → Missile → Photon Torpedo).
 */
export class UpgradeSystem {
  private readonly gameState: GameState;

  /**
   * Event fired when weapon research starts.
   * Parameters: (faction, weaponTier)
   */
  public onResearchStarted?: (faction: FactionType, weaponTier: WeaponTier) => void;

  /**
   * Event fired when weapon research completes.
   * Parameters: (faction, weaponTier)
   */
  public onResearchCompleted?: (faction: FactionType, weaponTier: WeaponTier) => void;

  // Research state per faction
  private playerWeaponTier: WeaponTier = WeaponTier.Laser;
  private aiWeaponTier: WeaponTier = WeaponTier.Laser;

  private playerResearchingTier: WeaponTier | null = null;
  private aiResearchingTier: WeaponTier | null = null;

  private playerResearchTurnsRemaining: number = 0;
  private aiResearchTurnsRemaining: number = 0;

  constructor(gameState: GameState) {
    if (!gameState) {
      throw new Error('gameState cannot be null or undefined');
    }

    this.gameState = gameState;
  }

  /**
   * Gets the current weapon tier for a faction.
   */
  public getWeaponTier(faction: FactionType): WeaponTier {
    return faction === FactionType.Player ? this.playerWeaponTier : this.aiWeaponTier;
  }

  /**
   * Checks if a faction is currently researching a weapon upgrade.
   */
  public isResearching(faction: FactionType): boolean {
    return faction === FactionType.Player
      ? this.playerResearchingTier !== null
      : this.aiResearchingTier !== null;
  }

  /**
   * Gets the weapon tier currently being researched (null if none).
   */
  public getResearchingTier(faction: FactionType): WeaponTier | null {
    return faction === FactionType.Player ? this.playerResearchingTier : this.aiResearchingTier;
  }

  /**
   * Gets the remaining research turns for a faction.
   */
  public getResearchTurnsRemaining(faction: FactionType): number {
    return faction === FactionType.Player
      ? this.playerResearchTurnsRemaining
      : this.aiResearchTurnsRemaining;
  }

  /**
   * Starts weapon upgrade research for a faction.
   * @param faction Faction researching
   * @returns True if research started, false if failed
   */
  public startResearch(faction: FactionType): boolean {
    // Get current and target tiers
    const currentTier = this.getWeaponTier(faction);
    const nextTier = this.getNextTier(currentTier);

    // Cannot research if already at max tier
    if (nextTier === null) {
      return false; // Already at Photon Torpedo
    }

    // Cannot research if already researching
    if (this.isResearching(faction)) {
      return false; // Research already in progress
    }

    // Get cost
    const cost = UpgradeCosts.getResearchCost(nextTier);

    // Validate resources (from faction resources)
    const factionState = faction === FactionType.Player ? this.gameState.playerFaction : this.gameState.aiFaction;
    if (factionState.resources.credits < cost) {
      return false; // Insufficient credits
    }

    // Deduct cost
    factionState.resources.credits -= cost;

    // Start research
    const researchTime = UpgradeCosts.getResearchTime(nextTier);

    if (faction === FactionType.Player) {
      this.playerResearchingTier = nextTier;
      this.playerResearchTurnsRemaining = researchTime;
    } else {
      this.aiResearchingTier = nextTier;
      this.aiResearchTurnsRemaining = researchTime;
    }

    this.onResearchStarted?.(faction, nextTier);
    return true;
  }

  /**
   * Updates research progress (called each turn during Income Phase).
   */
  public updateResearch(): void {
    // Update player research
    if (this.playerResearchingTier !== null && this.playerResearchTurnsRemaining > 0) {
      this.playerResearchTurnsRemaining--;

      if (this.playerResearchTurnsRemaining <= 0) {
        this.completeResearch(FactionType.Player);
      }
    }

    // Update AI research
    if (this.aiResearchingTier !== null && this.aiResearchTurnsRemaining > 0) {
      this.aiResearchTurnsRemaining--;

      if (this.aiResearchTurnsRemaining <= 0) {
        this.completeResearch(FactionType.AI);
      }
    }
  }

  /**
   * Completes weapon research and upgrades all Battle Cruisers.
   */
  private completeResearch(faction: FactionType): void {
    const researchedTier = faction === FactionType.Player ? this.playerResearchingTier : this.aiResearchingTier;

    if (!researchedTier) {
      return;
    }

    // Upgrade faction weapon tier
    if (faction === FactionType.Player) {
      this.playerWeaponTier = researchedTier;
      this.playerResearchingTier = null;
      this.playerResearchTurnsRemaining = 0;
    } else {
      this.aiWeaponTier = researchedTier;
      this.aiResearchingTier = null;
      this.aiResearchTurnsRemaining = 0;
    }

    // Get all Battle Cruisers owned by this faction for future upgrade
    const battleCruisers = this.gameState.craft.filter(
      c => c.owner === faction && c.type === CraftType.BattleCruiser
    );

    // Future combat system will apply weapon tier to these craft
    // For now, log the count and track the faction's weapon tier
    if (battleCruisers.length > 0) {
      // Prepared for future weapon tier application
    }

    this.onResearchCompleted?.(faction, researchedTier);
  }

  /**
   * Gets the next weapon tier in progression.
   */
  private getNextTier(current: WeaponTier): WeaponTier | null {
    switch (current) {
      case WeaponTier.Laser:
        return WeaponTier.Missile;
      case WeaponTier.Missile:
        return WeaponTier.PhotonTorpedo;
      case WeaponTier.PhotonTorpedo:
        return null; // Max tier
      default:
        return null;
    }
  }

  /**
   * Checks if a faction can afford to start the next weapon research.
   */
  public canAffordNextResearch(faction: FactionType): boolean {
    const currentTier = this.getWeaponTier(faction);
    const nextTier = this.getNextTier(currentTier);

    if (nextTier === null) {
      return false; // Already at max tier
    }

    const cost = UpgradeCosts.getResearchCost(nextTier);
    const factionState = faction === FactionType.Player ? this.gameState.playerFaction : this.gameState.aiFaction;

    return factionState.resources.credits >= cost;
  }

  /**
   * Gets the damage modifier for a faction's current weapon tier.
   */
  public getDamageModifier(faction: FactionType): number {
    const tier = this.getWeaponTier(faction);
    return UpgradeCosts.getDamageModifier(tier);
  }
}
